model/src/main/kotlin/utils/FindingsMatcher.kt

```diff
@@ -157,18 +152,38 @@ class FindingsMatcher(
             unmatchedCopyrights += copyrights.toSet() - matchedFileFindings.values.flatten()
         }
 
-        val rootLicenseFindings = getRootLicenseFindings(licenseFindings)
+        val matchedRootLicenseFindings = matchWithRootLicenses(licenseFindings, unmatchedCopyrights)
 
-        return if (rootLicenseFindings.isEmpty()) {
-            FindingsMatcherResult(matchedFindings, unmatchedCopyrights)
-        } else {
-            // Associate all unmatched copyright findings with all root licenses findings.
-            rootLicenseFindings.forEach { licenseFinding ->
-                matchedFindings[licenseFinding] = matchedFindings.getValue(licenseFinding) + unmatchedCopyrights
-            }
+        return FindingsMatcherResult(
+            matchedFindings = matchedFindings + matchedRootLicenseFindings,
+            unmatchedCopyrights = unmatchedCopyrights - matchedRootLicenseFindings.values.flatten()
+        )
+    }
+
+    /**
+     * Associate the given [copyrightFindings] to its corresponding applicable root licenses. If no root license is
+     * applicable to a given copyright finding, that copyright finding is not contained in the result.
+     */
+    fun matchWithRootLicenses(
+        licenseFindings: Set<LicenseFinding>,
+        copyrightFindings: Set<CopyrightFinding>
+    ): Map<LicenseFinding, Set<CopyrightFinding>> {
+        fun TextLocation.directory(): String = path.substringBeforeLast(delimiter = "/", missingDelimiterValue = "")
```

Comment:
This function is duplicated with the function at the bottom of the file.

Reply:
Done.